.github/workflows/rust.yml

```diff
@@ -79,28 +82,68 @@ jobs:
-        run: cargo check --profile ci --no-default-features -p datafusion
-
-      - name: Check datafusion-common without default features
+      - name: Check datafusion-common (no-default-features)
         run: cargo check --profile ci --all-targets --no-default-features -p datafusion-common
+        # Note: don't check other feature flags as datafusion-common is not typically used standalone
 
-      - name: Check datafusion-functions without default features
-        run: cargo check --profile ci --all-targets --no-default-features -p datafusion-functions
-
-      - name: Check datafusion-substrait without default features
+  # Check datafusion-substrait features
+  #
+  # Ensure via `cargo check` that the crate can be built with a
+  # subset of the features packages enabled.
+  linux-datafusion-substrait-features:
+    name: cargo check datafusion-substrait features
+    needs: linux-build-lib
+    runs-on: ubuntu-latest
+    container:
+      image: amd64/rust
+    steps:
+      - uses: actions/checkout@v4
+      - name: Setup Rust toolchain
+        uses: ./.github/actions/setup-builder
+        with:
+          rust-version: stable
+      - name: Check datafusion-substrait (no-default-features)
         run: cargo check --profile ci --all-targets --no-default-features -p datafusion-substrait
+      - name: Check datafusion-substrait (physical)
+        run: cargo check --profile ci --all-targets --no-default-features -p datafusion-substrait --features=physical
+      - name: Install Protobuf Compiler
+        run: |
+          apt-get update
+          apt-get install -y protobuf-compiler
+      - name: Check datafusion-substrait (protoc)
+        run: cargo check --profile ci --all-targets --no-default-features -p datafusion-substrait --features=protoc
 
-      - name: Check workspace in debug mode
-        run: cargo check --profile ci --all-targets --workspace
-
-      - name: Check workspace with additional features
-        run: cargo check --profile ci --workspace --benches --features avro,json,integration-tests
-
-  # cargo check datafusion to ensure that the datafusion crate can be built with only a
-  # subset of the function packages enabled.
+  # Check datafusion-proto features
+  #
+  # Ensure via `cargo check` that the crate can be built with a
+  # subset of the features packages enabled.
+  linux-datafusion-proto-features:
+    name: cargo check datafusion-proto features
+    needs: linux-build-lib
+    runs-on: ubuntu-latest
+    container:
+      image: amd64/rust
+    steps:
+      - uses: actions/checkout@v4
+      - name: Setup Rust toolchain
+        uses: ./.github/actions/setup-builder
+        with:
+          rust-version: stable
+      - name: Check datafusion-proto (no-default-features)
+        run: cargo check --profile ci --all-targets --no-default-features -p datafusion-proto
+      # fails due to https://github.com/apache/datafusion/issues/15157
```

Comment:
This adds coverage for the datafusion-proto crate, and in fact found a bug:
- https://github.com/apache/datafusion/issues/15157